runtime/src/impls.rs

```diff
@@ -123,6 +128,122 @@ impl Convert<(Weight, Fixed64), Fixed64> for FeeMultiplierUpdateHandler {
+
+		if let None = fee_exchange {
+			// User will pay for gas in CENNZnet's native fee currency
+			let imbalance = T::Currency::withdraw(
+				transactor,
+				fill_meter_cost,
+				WithdrawReason::Fee.into(),
+				ExistenceRequirement::KeepAlive,
+			)?;
+			T::GasPayment::on_unbalanced(imbalance);
+			return Ok(GasMeter::with_limit(gas_limit, gas_price));
+		}
+
+		// User wants to pay fee in a nominated currency
+		let exchange_op = fee_exchange.unwrap();
+		let payment_asset = exchange_op.asset_id();
+
+		// Calculate the `fill_meter_cost` in terms of the user's nominated payment asset
+		let converted_fill_meter_cost = CennzxSpot::<T>::get_asset_to_core_output_price(
+			&payment_asset,
+			T::Balance::unique_saturated_from(fill_meter_cost.saturated_into()),
+			CennzxSpot::<T>::fee_rate(),
+		)?;
+
+		// Respect the user's max. fee preference
+		if converted_fill_meter_cost > exchange_op.max_payment() {
+			return Err("Fee cost exceeds max. payment limit".into());
+		}
+
+		// Does the user have enough funds to pay the `converted_fill_meter_cost` with `payment_asset`
+		// also taking into consideration any liquidity restrictions
+		GenericAsset::<T>::ensure_can_withdraw(
+			&payment_asset,
+			transactor,
+			converted_fill_meter_cost,
+			WithdrawReason::Fee.into(),
+			// Calculate the expected user balance after paying the `converted_fill_meter_cost`
+			// This value is required to ensure liquidity restrictions are upheld
+			GenericAsset::<T>::free_balance(&payment_asset, transactor)
+				.checked_sub(&converted_fill_meter_cost)
+				.ok_or("Insufficient liquidity to fill gas meter")?,
+		)?;
+
+		// User has the requisite amount of `payment_asset` to fund the meter
+		// Actual payment will be handled in `empty_unused_gas` as the user may not spend the entire limit
+		// Performing payment on the known gas spent will avoid a refund situation
+		return Ok(GasMeter::with_limit(gas_limit, gas_price));
+	}
+
+	/// Handle settlement of unused gas after contract execution
+	///
+	/// The process is as follows:
+	/// - Default case: refund unused gas tokens to the user (`transactor`) in CENNZnet's native fee currency as the current gas price
+	/// - FeeExchange case: Gas spent will be charged to the user in their nominated fee currency at the current gas price
+	fn empty_unused_gas(transactor: &T::AccountId, gas_meter: GasMeter<T>) {
+		// TODO: Update `GasSpent` for the block
+		let gas_left = gas_meter.gas_left();
+		let gas_price = Contracts::<T>::gas_price();
+		let gas_spent = gas_meter.spent();
+
+		// The `take()` function ensures the entry is killed after access
+		if let Some(exchange_op) = storage::unhashed::take::<FeeExchange<T::AssetId, T::Balance>>(&GAS_FEE_EXCHANGE_KEY)
+		{
+			// Pay for `gas_spent` in a user nominated currency using the CENNZX spot exchange
+			// Payment can never fail as liquidity is verified before filling the meter
+			if let Some(used_gas_cost) = gas_price.checked_mul(&gas_spent.saturated_into()) {
```

Comment:
Thank you for spotting that the gas spent hadn't been multiplied by the gas price here.